x-pack/plugins/apm/public/hooks/table_sort_pagination/use_table_sort_pagination.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  CriteriaWithPagination,
+  EuiTableSortingType,
+  Pagination,
+} from '@elastic/eui';
+import { orderBy } from 'lodash';
+import { useMemo } from 'react';
+import uuid from 'uuid';
+import { useApmParams } from '../use_apm_params';
+
+export interface TableSortPaginationProps<T extends any[]> {
+  items: T;
+  // totalItemCount is not necessary here
+  initialPagination: Omit<Pagination, 'totalItemCount'>;
+  initialSort?: EuiTableSortingType<T[0]>;
+}
+
+type Props<T extends any[]> = TableSortPaginationProps<T> & {
+  tableOptions: CriteriaWithPagination<T[0]>;
+  onTableChange: (criteriaWithPagination: CriteriaWithPagination<T[0]>) => void;
+};
+
+const PAGE_SIZE_OPTIONS = [10, 25, 50];
+
+export function useTableSortAndPagination<T extends any[]>({
+  items,
+  initialPagination,
+  initialSort = {},
+  tableOptions,
+  onTableChange,
+}: Props<T>): {
+  onTableChange: (criteriaWithPagination: CriteriaWithPagination<T[0]>) => void;
+  tableSort?: EuiTableSortingType<T[0]>;
+  tablePagination: Pagination;
+  tableItems: any[];
+  totalItems: number;
+  requestId: string;
+} {
+  const { query } = useApmParams('/*');
+  const offset = 'offset' in query ? query.offset : undefined;
+  const comparisonEnabled =
+    'comparisonEnabled' in query ? query.comparisonEnabled : undefined;
+
+  const { tableItems, requestId } = useMemo(
+    () => {
+      return {
+        tableItems: orderBy(
+          items,
+          tableOptions.sort?.field,
+          tableOptions.sort?.direction
+        ).slice(
+          tableOptions.page.index * tableOptions.page.size,
+          (tableOptions.page.index + 1) * tableOptions.page.size
+        ),
+        // Generate a new id everytime the table options are changed
+        requestId: uuid(),
+      };
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [
+      items,
```

Comment:
This seems rather brittle, this could easily invalidate the dependencies of the hook if someone transforms `items`, e.g. with a `.map()`.

Reply:
But that is kind of necessary, right? since I need to update the `tableItems` when new Items come in if, for example, I push the refresh button?